packages/router/contracts/adapters/PendleAdapter.sol

```diff
@@ -0,0 +1,387 @@
+      // pt redeem -> sy to uniswap -> tokenOut to recipient
+      IMarginlyRouter(msg.sender).adapterCallback(address(this), amountIn, data);
+
+      uint256 syAmountOut = marketData.yt.redeemPY(marketData.uniswapV3);
+      // TODO is this even possible?
+      if (syAmountOut != amountIn) revert();
+
+      (, amountOut) = _uniswapV3LikeSwap(
+        recipient,
+        marketData.uniswapV3,
+        address(marketData.sy) < tokenOut,
+        int256(syAmountOut),
+        ''
+      );
+    } else if (tokenOut == address(marketData.pt)) {
+      // sy to pt swap is not possible after maturity
+      revert Forbidden();
+    } else {
+      revert();
+    }
+  }
+
+  function _swapExactOutputPostMaturity(
+    PendleMarketData memory marketData,
+    address recipient,
+    address tokenIn,
+    address tokenOut,
+    uint256 amountOut,
+    uint256 maxAmountIn,
+    bytes calldata data
+  ) private returns (uint256 amountIn) {
+    CallbackData memory swapCallbackData = CallbackData({
+      tokenIn: tokenIn,
+      tokenOut: tokenOut,
+      syToken: address(marketData.sy),
+      initiator: msg.sender,
+      maxAmountIn: maxAmountIn,
+      data: data
+    });
+    if (tokenIn == address(marketData.pt)) {
+      // call pendle -> swapCallback triggers uniswapV3 tokenIn/sy swap -> completing pendle swap
+      (amountIn, ) = marketData.market.swapSyForExactPt(recipient, amountOut, abi.encode(swapCallbackData));
+    } else if (tokenOut == address(marketData.pt)) {
+      // sy to pt swap is not possible after maturity
+      revert Forbidden();
+    } else {
+      revert();
+    }
+  }
+
+  function _uniswapV3LikeSwap(
+    address recipient,
+    address pool,
+    bool zeroForOne,
+    int256 swapAmount,
+    bytes memory data
+  ) internal returns (uint256 amountIn, uint256 amountOut) {
+    (int256 amount0Delta, int256 amount1Delta) = IUniswapV3Pool(pool).swap(
+      recipient,
+      zeroForOne,
+      swapAmount,
+      zeroForOne ? MIN_SQRT_RATIO + 1 : MAX_SQRT_RATIO - 1,
+      data
+    );
+
+    (amountIn, amountOut) = zeroForOne
+      ? (uint256(amount0Delta), uint256(-amount1Delta))
+      : (uint256(amount1Delta), uint256(-amount0Delta));
+  }
+
+  function _pendleApproxSwapExactSyForPt(
+    PendleMarketData memory marketData,
+    address recipient,
+    uint256 syAmountIn,
+    uint256 minPtAmountOut
+  ) private returns (uint256 ptAmountOut) {
+    ApproxParams memory approx = ApproxParams({
+      guessMin: minPtAmountOut,
+      guessMax: (minPtAmountOut * ONE_PLUS_SLIPPAGE) / ONE_MINUS_SLIPPAGE,
+      guessOffchain: 0,
+      maxIteration: MAX_ITERATIONS,
+      eps: EPSILON
+    });
+
+    (ptAmountOut, ) = MarketApproxPtOutLib.approxSwapExactSyForPt(
+      marketData.market.readState(address(this)),
+      marketData.yt.newIndex(),
+      syAmountIn,
+      block.timestamp,
+      approx
+    );
+
+    (uint256 actualAmountIn, ) = marketData.market.swapSyForExactPt(recipient, ptAmountOut, '');
```

Comment:
I use the following names of token types:
- SY - Standardized Yield (name from Pendle)
- PT - Principal Token (name from Pendle)
- YQT - Yield Quote Token (suggested name)
- QT - Quote Token (suggested name)

Pools contains tokens:
Pendle pool: PT/SY
External additional pool: YQT/QT

Exchange rate: 1.0 SY = 1.0 YQT

Does `swapSyForExactPt` work with YQT token? At this moment adapter has only YQT, not SY